File: CS/E4121/Models/LargeDatabaseDataProvider.cs
using DevExpress.Data.Filtering;
using DevExpress.Data.Linq;
using DevExpress.Data.Linq.Helpers;
using System.Collections.Generic;
using System.Linq;

namespace E4121.Models
{
    public static class LargeDatabaseDataProvider
    {
        static LargeDatabaseDataContext db;
        public static LargeDatabaseDataContext DB
        {
            get
            {
                if (db == null)
                    db = new LargeDatabaseDataContext();
                return db;
            }
        }

        public static IList<Email> GetEmailsByFilter(CriteriaOperator filterCriteria)
        {
            CriteriaToExpressionConverter converter = new CriteriaToExpressionConverter();

            IQueryable<Email> source = from product in DB.Emails
                                       select product;
            IQueryable<Email> filteredData = source.AppendWhere(converter, filterCriteria) as IQueryable<Email>;

            return filteredData.ToList<Email>();
        }
    }
}

File: VB/E4121VB/Models/LargeDatabaseDataProvider.vb
Imports DevExpress.Data.Filtering
Imports DevExpress.Data.Linq
Imports DevExpress.Data.Linq.Helpers

Public Module LargeDatabaseDataProvider
    Dim _db As LargeDatabaseDataContext

    Public ReadOnly Property DB As LargeDatabaseDataContext
        Get
            If _db Is Nothing Then _db = New LargeDatabaseDataContext()
            Return _db
        End Get
    End Property

    Function GetEmailsByFilter(ByVal filterCriteria As CriteriaOperator) As IList(Of Email)
        Dim converter As CriteriaToExpressionConverter = New CriteriaToExpressionConverter()
        Dim source As IQueryable(Of Email) = From product In DB.Emails Select product
        Dim filteredData As IQueryable(Of Email) = TryCast(source.AppendWhere(converter, filterCriteria), IQueryable(Of Email))
        Return filteredData.ToList()
    End Function
End Module
